sympy/solvers/solveset.py

```diff
@@ -871,3 +872,162 @@ def solveset(f, symbol=None):
+
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.solveset import eq_to_matrix
+    >>> from sympy import symbols
+    >>> x, y, z = symbols('x, y, z')
+    >>> eqns = [x + 2*y + 3*z - 1, 3*x + y + z + 6, 2*x + 4*y + 9*z - 2]
+    >>> M = eq_to_matrix(eqns, [x, y, z])
+    >>> M
+    Matrix([
+    [1, 2, 3,  1],
+    [3, 1, 1, -6],
+    [2, 4, 9,  2]])
+
+    >>> eqns = [x + z - 1, y + z, x - y]
+    >>> M = eq_to_matrix(eqns, [x, y, z])
+    >>> M
+    Matrix([
+    [1,  0, 1,  1],
+    [0,  1, 1,  0],
+    [1, -1, 0,  0]])
+
+
+    """
+
+    if hasattr(symbols[0], '__iter__'):
+        symbols = symbols[0]
+
+    M = Matrix([symbols])
+    # initialise Matrix with symbols + 1 columns
+    M = M.col_insert(len(symbols), Matrix([1]))
+    row_no = 1
+
+    for equation in equations:
+        f = sympify(equation)
+        if isinstance(f, Equality):
+            f = f.lhs - f.rhs
+
+        # Extract coeff of symbols
+        coeff_list = []
+        for symbol in symbols:
+            coeff_list.append(f.coeff(symbol))
+
+        # append constant term (term free from symbols)
+        coeff_list.append(-f.as_coeff_add(*symbols)[0])
+
+        # insert equations coeff's into rows
+        M = M.row_insert(row_no, Matrix([coeff_list]))
+        row_no += 1
+
+    # delete the initialised (Ist) trivial row
+    M.row_del(0)
+    return M
+
+
+def linsolve(system, *symbols):
+    """ Solve system of N linear equations with M variables.
+
+    'EXPERIMENTAL CODE PUSHED CURRENTLY'
+
+    All Standard input formats are supported:
+    For the given set of Equations, the respective input types are given below:
+
+    3*x + 2*y -   z = 1
+    2*x - 2*y + 4*z = -2
+    2*x -   y + 2*z = 0
+
+    * Augmented Matrix Form
+
+                 [3   2  -1  1]
+    system   =   [2  -2   4 -2]  (Matrix)
+                 [2  -1   2  0]
+
+    * List Of Equations Form
+
+    system  =  [3*x + 2*y - z - 1, 2*x - 2*y + 4*z + 2, 2*x - y + 2*z]
+
+    * Input A & b Matrix Form (from Ax = b)
+
+            [3   2  -1 ]          [  1 ]
+    A   =   [2  -2   4 ]    b  =  [ -2 ]
+            [2  -1   2 ]          [  0 ]
+
+    system = (A, b)
+
+    Examples
+    ========
+
+    TODO
+
+    """
+
+    # 1). Augmented Matrix as input
+    if isinstance(system, Matrix):
+        aug = system
+
+    elif hasattr(system, '__iter__'):
+        if isinstance(system[0], Matrix):
+            # 2). A & b as input
+            if len(system) == 2:
+                A = system[0]
+                b = system[1]
+                aug = A.hstack(A.copy(), b.copy())
+            else:
+                raise ValueError("For tuple as first argument expected length \
+                                 is 2: (A, b)")
+
+        else:
+            # 3). List of equations Form
+            aug = eq_to_matrix(system, symbols)
+    else:
+        raise ValueError("Invalid arguments")
+
+    m, n = aug[:, :-1].shape
+
+    # solve by row reduced echleon form using Gauss Elimination
```

Comment:
btw, I don't intend to implement Least squares solution, since they are approximations (numerical stuff), by under-determined I mean returning parametric solutions, like say:  `{(y-2z, y, z)}` .